### PR TITLE
Paypal redirect message

### PR DESCRIPTION
**Pull request by sktoiva**
This pull request show a redirection message and a manual link to user trying to connect their PayPal account.

We've had cases where the redirection step fails, and they are hard to debug as the user doesn't understand what goes wrong. This fixes it by showing more information.

### CODE REVIEW

File: app/assets/javascripts/paypal_account_settings.js [hunk 1]
```diff
@@ -2,23 +2,28 @@ window.ST = window.ST || {};
 
 (function(module) {
 
-  module.initializePayPalAccountForm = function(formId) {
-    var form = $('#'+formId);
+
+  module.initializeNewPayPalAccountHandler = function(button_id, action) {
```
Comment (rap1ds) [inline app/assets/javascripts/paypal_account_settings.js]: Too much Ruby makes you write snake_cases in JavaScript, although the way to go in JS is camelCase

File: app/assets/javascripts/paypal_account_settings.js [hunk 2]
```diff
@@ -2,23 +2,28 @@ window.ST = window.ST || {};
...
-    form.validate({
-      submitHandler: function(form) {
-        var $form = $(form);
-        var $sendButton = $form.find(".send_button");
-        if(!$sendButton.hasClass("send-button-loading")) {
-          $form.find(".send-button-wrapper").append(spinner);
-          $sendButton.addClass("send-button-loading").blur();
-          form.submit();
+    $button.click(function(){
+      var $button_wrapper = $button.parent()
```
Comment (rap1ds) [inline app/assets/javascripts/paypal_account_settings.js]: Same here, camelCase instead

File: app/controllers/paypal_accounts_controller.rb [hunk 3]
```diff
@@ -101,7 +101,8 @@ def ask_order_permission
       flash[:error] = t("paypal_accounts.new.could_not_fetch_redirect_url")
       return redirect_to action: :new
     else
-      return redirect_to permissions_url
+      render json: {redirect_url: permissions_url,
+                    redirect_message: "#{t("paypal_accounts.redirect_message")} <a href=#{permissions_url}>#{t("paypal_accounts.redirect_here")}</a>"}
```
Comment (rap1ds) [inline app/controllers/paypal_accounts_controller.rb]: I feel that we shouldn't send HTML in the JSON content. Could we send the redirect_url (as we are actually already sending it) and the translation separately, and then construct the HTML in the view (JS) layer?

File: app/controllers/paypal_accounts_controller.rb [hunk 4]
```diff
@@ -134,7 +135,8 @@ def ask_billing_agreement
         flash[:error] = t("paypal_accounts.new.could_not_fetch_redirect_url")
         return redirect_to action: :new
       else
-        return redirect_to billing_agreement_url
+        render json: {redirect_url: billing_agreement_url,
+                      redirect_message: "#{t("paypal_accounts.redirect_message")} <a href=#{billing_agreement_url}>#{t("paypal_accounts.redirect_here")}</a>"}
```
Comment (rap1ds) [inline app/controllers/paypal_accounts_controller.rb]: Same here

File: app/views/paypal_accounts/_ask_paypal_billing_agreement.haml [hunk 5]
```diff
@@ -1,14 +1,18 @@
...
+
 .row.paypal-connected-email-row
   %h3.paypal-account-connected
     =t("paypal_accounts.paypal_account_email_connected").html_safe
   .paypal-connected-email.paypal-account-connected
     = t("paypal_accounts.connected_account", :email => paypal_account_email)
-  .paypal-connected-change
-    = link_to t("paypal_accounts.change_account"), change_url
+  .paypal-connected-change.send-button-wrapper
+    = content_tag(:button, t("paypal_accounts.change_account"), type: :button, id: "ask_paypal_permissions")
```
Comment (rap1ds) [inline app/views/paypal_accounts/_ask_paypal_billing_agreement.haml]: `content_tag` is new to me. Haven't seen it before.

What's the advantage vs using?:

```
%button#ask_paypal_permissions
  = t("paypal_accounts.change_account")
```

Comment (sktoiva) [inline app/views/paypal_accounts/_ask_paypal_billing_agreement.haml]: None, just long time not writing haml :)

File: app/assets/javascripts/paypal_account_settings.js [hunk 6]
```diff
@@ -2,23 +2,28 @@ window.ST = window.ST || {};
...
-  module.initializePayPalAccountForm = function(formId) {
-    var form = $('#'+formId);
+
+  module.initializeNewPayPalAccountHandler = function(button_id, action) {
+    var $button = $('#'+button_id);
     var spinner = new Image();
     spinner.src = "https://s3.amazonaws.com/sharetribe/assets/ajax-loader-grey.gif";
     spinner.className = "send-button-loading-img";
 
-    form.validate({
```
Comment (rap1ds) [inline app/assets/javascripts/paypal_account_settings.js]: What the heck, I don't understand this... There was a form validator, but it didn't have any validations configured, only the submit handler?

File: app/assets/javascripts/paypal_account_settings.js (hunk 1 above)
Comment (rap1ds) [inline app/assets/javascripts/paypal_account_settings.js]: Btw. in the code we have always used `Paypal`, which is wrong, yes, but for the sake of consistency, should we use `Paypal` also here?

**Comment by rap1ds** [comment]
Check JSHint, there are some missing semicolons

**Comment by rap1ds** [comment]
Did you test the admin side?

I got an error while trying to connect admin PayPal:

```
No route matches {:action=>"ask_order_permission", :controller=>"paypal_accounts", :locale=>:en}
```

**Comment by vsaarinen** [comment]
Copy reviewed.

File: app/assets/javascripts/paypal_account_settings.js (hunk 6 above)
Comment (sktoiva) [inline app/assets/javascripts/paypal_account_settings.js]: Yes, that was the reason I refactored it out. I couldn't understand why there was a form in the first place. Maybe legacy from the admin side or something.

File: app/views/admin/paypal_preferences/index.haml [hunk 7]
```diff
@@ -33,18 +33,21 @@
...
-        ST.initializePayPalAccountForm("paypal_account_form");
+      .paypal-connected-change.send-button-wrapper
+        - content_for :extra_javascript do
+          :javascript
+            ST.initializeNewPaypalAccountHandler("ask_paypal_permissions", "#{order_permission_action}");
+        %button#ask_paypal_permissions
+          = t("paypal_accounts.change_account")
+    .row
+      .paypal-redirect-message.hidden
+        - paypal_redirect_link = "<a href='#' id='ask_paypal_permissions_redirect'>#{t("paypal_accounts.redirect_link_text")}</a>"
```
Comment (rap1ds) [inline app/views/admin/paypal_preferences/index.haml]: You could also use normal HAML and `link_to` helper here

File: app/views/paypal_accounts/_ask_paypal_billing_agreement.haml [hunk 8]
```diff
@@ -1,14 +1,23 @@
...
   .paypal-connected-email.paypal-account-connected
     = t("paypal_accounts.connected_account", :email => paypal_account_email)
-  .paypal-connected-change
-    = link_to t("paypal_accounts.change_account"), change_url
+  .paypal-connected-change.send-button-wrapper
+    %button#ask_paypal_permissions
+      = t("paypal_accounts.change_account")
+.row
+  .paypal-redirect-message.hidden
+    - paypal_redirect_link = "<a href='#' id='ask_paypal_permissions_redirect'>#{t("paypal_accounts.redirect_link_text")}</a>"
```
Comment (rap1ds) [inline app/views/paypal_accounts/_ask_paypal_billing_agreement.haml]: Same here

File: app/views/paypal_accounts/_ask_paypal_billing_agreement.haml [hunk 9]
```diff
@@ -27,7 +36,12 @@
       paypal_info_link: paypal_info_link).html_safe }
 
   .row.send-button-wrapper
-    = form.button t("paypal_accounts.new.billing_agreement"), :class => "send_button"
+    %button.send_button#ask_billing_agreement
+      = t("paypal_accounts.new.billing_agreement")
+  .row
+    .paypal-redirect-message.hidden
+      - paypal_redirect_link = "<a href='#' id='ask_billing_agreement_redirect'>#{t("paypal_accounts.redirect_link_text")}</a>"
```
Comment (rap1ds) [inline app/views/paypal_accounts/_ask_paypal_billing_agreement.haml]: And here

File: app/views/paypal_accounts/_ask_paypal_permissions.haml [hunk 10]
```diff
@@ -7,4 +11,9 @@
   = render :partial => "layouts/info_text", :locals => {text: text}
 
 .row.send-button-wrapper
-  = form.button t("paypal_accounts.new.connect_paypal_account"), :class => "send_button"
+  %button#ask_paypal_permissions
+    = t("paypal_accounts.new.connect_paypal_account")
+.row
+  .paypal-redirect-message.hidden
+    - paypal_redirect_link = "<a href='#' id='ask_paypal_permissions_redirect'>#{t("paypal_accounts.redirect_link_text")}</a>"
```
Comment (rap1ds) [inline app/views/paypal_accounts/_ask_paypal_permissions.haml]: And here

File: app/views/paypal_accounts/show.haml [hunk 11]
```diff
@@ -15,8 +18,13 @@
...
       .paypal-connected-email.paypal-account-connected
         = t("paypal_accounts.connected_account", :email => paypal_account_email)
-      .paypal-connected-change
-        = link_to t("paypal_accounts.change_account"), change_url
+      .paypal-connected-change.send-button-wrapper
+        %button#ask_paypal_permissions
+          = t("paypal_accounts.change_account")
+    .row
+      .paypal-redirect-message.hidden
+        - paypal_redirect_link = "<a href='#' id='ask_paypal_permissions_redirect'>#{t("paypal_accounts.redirect_link_text")}</a>"
```
Comment (rap1ds) [inline app/views/paypal_accounts/show.haml]: And here

File: app/assets/javascripts/paypal_account_settings.js [hunk 12]
```diff
@@ -2,22 +2,29 @@ window.ST = window.ST || {};
...
+      $buttonWrapper.append(spinner);
+      $button.addClass("send-button-loading").blur();
+
+      $.ajax({
+        type: 'GET',
+        url: action,
+        success: function(response){
+          var $redirectLink = $('#' + buttonId + '_redirect');
+          $redirectLink.attr('href', response.redirect_url);
+          $redirectLink.parent().toggleClass('hidden');
```
Comment (rap1ds) [inline app/assets/javascripts/paypal_account_settings.js]: Using `parent` makes the code a bit fragile. Now the code is tightly coupled to the DOM structure and may break if someone for example adds another innocent looking wrapper div around the link.

**Comment by rap1ds** [comment]
Tested but personal and admin side. Works nicely!

Added few comments but they are not that important. Good to go imo.

**Pull request merged**